File: packages/java/src/main/java/com/ailabtools/sdk/http/FileInput.java
package com.ailabtools.sdk.http;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Objects;

/** A file uploaded as part of a multipart API request. */
public final class FileInput {
    private final byte[] bytes;
    private final String filename;
    private final String contentType;

    private FileInput(byte[] bytes, String filename, String contentType) {
        this.bytes = Arrays.copyOf(bytes, bytes.length);
        this.filename = Objects.requireNonNull(filename, "filename");
        this.contentType = contentType == null ? detectContentType(filename) : contentType;
    }

    public static FileInput fromPath(Path path) throws IOException {
        return fromPath(path, path.getFileName().toString(), null);
    }

    public static FileInput fromPath(Path path, String filename, String contentType) throws IOException {
        return new FileInput(Files.readAllBytes(path), filename, contentType);
    }

    public static FileInput fromBytes(byte[] bytes, String filename) {
        return new FileInput(bytes, filename, null);
    }

    public static FileInput fromBytes(byte[] bytes, String filename, String contentType) {
        return new FileInput(bytes, filename, contentType);
    }

    public static FileInput fromInputStream(InputStream stream, String filename) throws IOException {
        Objects.requireNonNull(stream, "stream");
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[8192];
        int read;
        while ((read = stream.read(buffer)) != -1) {
            output.write(buffer, 0, read);
        }
        return new FileInput(output.toByteArray(), filename, null);
    }

    public byte[] bytes() {
        return Arrays.copyOf(bytes, bytes.length);
    }

    public String filename() {
        return filename;
    }

    public String contentType() {
        return contentType;
    }

    private static String detectContentType(String filename) {
        String lower = filename.toLowerCase();
        if (lower.endsWith(".jpg") || lower.endsWith(".jpeg")) return "image/jpeg";
        if (lower.endsWith(".png")) return "image/png";
        if (lower.endsWith(".webp")) return "image/webp";
        if (lower.endsWith(".bmp")) return "image/bmp";
        return "application/octet-stream";
    }
}

File: packages/kotlin/src/main/kotlin/com/ailabtools/sdk/FileInput.kt
package com.ailabtools.sdk

import java.io.File
import java.io.InputStream

/** Binary file content used by multipart API parameters. */
public class FileInput private constructor(
    internal val bytes: ByteArray,
    public val filename: String,
    public val contentType: String,
) {
    public companion object {
        @JvmStatic
        @JvmOverloads
        public fun fromBytes(
            bytes: ByteArray,
            filename: String,
            contentType: String = "application/octet-stream",
        ): FileInput {
            require(filename.isNotBlank()) { "filename must not be blank" }
            return FileInput(bytes.copyOf(), filename, contentType)
        }

        @JvmStatic
        @JvmOverloads
        public fun fromFile(
            file: File,
            contentType: String = "application/octet-stream",
        ): FileInput = fromBytes(file.readBytes(), file.name, contentType)

        @JvmStatic
        @JvmOverloads
        public fun fromStream(
            input: InputStream,
            filename: String,
            contentType: String = "application/octet-stream",
        ): FileInput = fromBytes(input.readBytes(), filename, contentType)
    }
}
